core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class for classification algorithms."""
+
+import collections
+import keyword
+import math
+import StringIO
+import token
+import tokenize
+
+from core.classifiers import base
+from core.classifiers.CodeClassifier import winnowing
+
+import numpy as np
+from sklearn.feature_extraction import text as sklearn_text
+from sklearn import model_selection
+from sklearn import svm
+
+# The string with which all the variable and method names need to be replaced.
+VAR_TOKEN = 'V'
+
+# The string with which will all unkown tokens (tokens which are ignored because
+# they appear rarely in a program) will be replaced.
+UNK_TOKEN = 'UNK'
+
+# T is the model parameter of winnowing algorithm. It is the minimum length of a
+# substring that winnowing should match when comparing two programs.
+# The minimum value of T.
+T_MIN = 3
+# The maximum value of T.
+T_MAX = 11
+
+# K is length of substring for which the hash value is to be generated. Each
+# substring of length K in program will be converted to equivalent hash
+# value (K <= T).
+# Minimum value of K (K <= T). K is model parameter of winnowing algorithm.
+K_MIN = 3
+
+# ID of the token which is a comment.
+COMMENT_TOKEN = 54
+
+# Threshold used when generating vocabulary for dataset. If token appears less
+# than threshold times then vocabulary ignores the token.
+VOCABULARY_THRESHOLD = 5
+
+
+def get_tokens(program):
+    """Generate tokens for program using tokenize module.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+
+    Yields:
+        int. Integer value which represents type of token.
+        str. Token.
+    """
+    for token_id, token_name, _, _, _ in tokenize.generate_tokens(
+            StringIO.StringIO(program).readline):
+        yield (token_id, token_name)
+
+
+def cv_tokenizer(program):
+    """Custom tokenizer for tokenizing Python programs.
+
+    This needs to be passed as an argument to sci-kit's CountVectorizer.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+        VAR_TOKEN: str. Replace variable and method names with this string when
+            tokenizing data.
+
+    Returns:
+        list(str). A list containing tokens of input program.
+    """
+    tokenized_program = []
+    for token_id, token_name in get_tokens(program):
+        # Ignore all newline, comment and empty string tokens.
+        # token_id is token.N_TOKENS for empty line / newline token.
+        # token_id is 54 for comments.
+        if (token_id == token.N_TOKENS or token_id == COMMENT_TOKEN
```

Comment:
Just FYI, note that tokenize.NL also exists: https://docs.python.org/2/library/tokenize.html#tokenize.NL Not sure if it's the same thing. I couldn't easily find a definition of N_TOKENS.